op-node/specs: refactor batch queue, add parent_hash to batches, extend testing

protolambda opened the pull request:
This adds `parent_hash` to the batch definition, to ensure we never skip a batch. Skipping a batch can cause transactions with invalid nonce values to be inserted into the execution engine, causing reorgs or complete halts.

The `parent_hash` matches the L2 block hash of the block the batch will be building on top of, i.e. the L2 safe head.
To enable this, we have to change the batch-queue processing to only output 1 batch at the time. Without this, the previous L2 block hash would not be available, and so we could not check the hash otherwise.

In a previous iteration of this I considered hashing just the batch inputs, and chaining those hashes. However, we can't include those hashes into the L2 chain elegantly, which is required to find a sync continuation point without loading all historical batches. I.e. we would need to add additional data to the L1 block info transaction, and it would almost be self-referencing, as it hashes together the other transactions, which are included later.
Using the native block hashes thus is preferred to not add additional data to the info deposit, and avoid the near cyclic dependency issue.

Not outputting multiple batches at a time really means that we are not filling multiple batches at a time when there is a gap to fill. The eager batch derivation already outputted a single batch otherwise.

Previously the validity conditions on batches were spread between the batch-filtering, batch-filling, extension-checking and batch queue main function. I changed this so we can output 1 batch at a time, check all validity conditions, and isolate the validity conditions in one function that now has 100% test coverage.

Part of this unwinding of the different checks also meant we can now remove the L1 RPC from the function: all necessary L1 data is there, and is easy to access (first L1 block always matches L2 safe head L1 origin, second L1 block is optional).

And validating the batches this way means there is no difference between eager and regular batch-derivation: we always consider the validity of the first batch, since we output one at a time.

The spec is updated to reflect this same methodology, so we don't need eager batch queue <> regular batch queue distinction anymore.

I hope the PR is not too large to review, this was not an easy change, at least it should reduce total lines of regular code, while adding more tests + ensuring 100% coverage of batch validation.

Fix ENG-2593

protolambda commented:
Rebased onto `develop` for the devnet CI fix.

norswap commented:
> The parent_hash matches the L2 block hash of the block the batch will be building on top of, i.e. the L2 safe head.
To enable this, we have to change the batch-queue processing to only output 1 batch at the time. Without this, the previous L2 block hash would not be available, and so we could not check the hash otherwise.

This has a pretty big consequence which is that if the sequencer does something bad on the black transition of batch `X`, then all the subsequent batches are rendered useless and cannot be replayed anymore. That's a pretty change and I feel like it should warrant some discussions of pros & cons.

> In a previous iteration of this I considered hashing just the batch inputs, and chaining those hashes. However, we can't include those hashes into the L2 chain elegantly, which is required to find a sync continuation point without loading all historical batches. I.e. we would need to add additional data to the L1 block info transaction, and it would almost be self-referencing, as it hashes together the other transactions, which are included later.

I'm not entirely sure why we can't include the hash of the previous batch in the L1 infos of the next block?

norswap commented on specs/derivation.md:
```diff
@@ -490,27 +491,63 @@ Note that the presence of any gaps in the batches derived from L1 means that thi
+- Otherwise,
+  - `epoch_num = next_epoch.number`
+  - `epoch_hash = next_epoch.hash`
 
```
I'm wondering if these spec changes are really necessary.

...

I no joke spent > 30 minutes writing about pros & cons of current & proposed spec, but truth be told, I'm frustrated with both.

I think what I'd love is for this to be rewritten more precisely in terms of:
- input: a stream of batches
- inner state: an ordered buffer + l2 safe head
- output: a stream of ordered batches (one for every slot)

and the transition function being what appens when  either (a) you process a batch from the input stream, or (b) the "L1 head" (i.e. latest fully processed L1 block) changes.

(btw I think ideally the whole spec should be structured like this and the implem should try to follow that format too)

I think for this format, the old spec will form a stronger base, because basically when a batch comes in you want to apply the timestamp + duplication filtering criteria from the old spec. Then you want to see if after adding the batch you have some contiguous batches at the start of your buffer, and if so push that to the next stage.

At that point you *haven't* applied the `parent_hash` or epoch number checks (though you probably could do the epoch number check by verifying that the epoch number does not go down as the timestamp go up?? maybe). These would be perfect to push to the next stage (e.g. only derive payload attributes when we have an up to date L2 safe head to build upon).

norswap reviewed (commented):
Didn't review the code really, mostly spec + general principles.

I think we need to talk about the implications of the block hash (see my comment separate from this review).

Specs-wise, I have some suggestions about how to better structure this bit of the spec. Maybe with impact to the code (new stage?)

Regarding the code, this is a lot of changes. I hear that it's quite hard to split, but it would be incredibly easy for something to slip through the cracks.

Wonder if we could have worked to temporary stages that are fully functional but a bit "frankenstein" in that they introduce code that will be removed in the very next PR? Not sure, but worth considering.

I think in this specific case, it would have been good to implement the change entirely without a refactor, then consider refactors separately.

And we could have split refactors in multiple PRs where each PR moves only a single check for one place to another.

Imho right now our options are basically: hope our testing catches any issues, or do the above. Don't feel very comfortable reviewing the changes and seeing "yup, I'm confident that before/after is the same behaviour modulo the new bit we're adding".

trianglesphere commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -244,76 +249,56 @@ func TestBatchQueueMissing(t *testing.T) {
-	// do eager batch processing for that batch. This test checks
-	// that batch timestamp 12 & 14 is created & 16 is used.
-	batches := []*BatchData{b(16, l1[0]), b(20, l1[1])}
+	// The batches at 18 and 20 are skipped to stop 22 from being eagerly processed.
```
really nice comments here.

trianglesphere commented on op-node/rollup/derive/batches.go:
```diff
@@ -1,106 +1,121 @@
-	if (batch.Timestamp-config.Genesis.L2Time)%config.BlockTime != 0 {
-		return fmt.Errorf("bad timestamp %d, not a multiple of the block time", batch.Timestamp)
+
+	if batch.Batch.Timestamp < nextTimestamp {
```
this plus `batch.Batch.Timestamp > nextTimestamp` is a weird way of saying
```
if batch.Batch.Timestamp != nextTimestamp {
    return ...
}
```

protolambda replied:
True, although the results are different: too old = drop, too new = buffer, exact equal = process.

trianglesphere reviewed (changes requested):
I went through and reviewed everything. I didn't realize how much you had refactored in there. Splitting this up into a refactor + adding parent hash as two steps (and maybe even smaller refactor steps depending on how it is) would have made this easier to review.

There's a couple spots that need some more comments.

My main concern is the repeated linear scan through `bq.batches` + how that array is modified. It does reduce the chance of leaking compared to the old design, but now requires (potentially) much more work.

protolambda commented on specs/derivation.md:
```diff
@@ -490,27 +491,63 @@ Note that the presence of any gaps in the batches derived from L1 means that thi
+- `transactions` is empty, i.e. no sequencer transactions. Deposited transactions may be added in the next stage.
+- If `next_timestamp < next_epoch.time`: the current L1 origin is repeated, to preserve the L2 time invariant.
+  - `epoch_num = epoch.number`
+  - `epoch_hash = epoch.hash`
```
That's exactly what we did previously: we can't ever break the time invariant or introduce a gap in L2 blocks, and must thus produce empty blocks with the old epoch, until we can meet the L1 time. And moving ahead to the next epoch is the right thing to do if we don't break the invariant, when we're forced to produce an empty block to move forward the deposit processing as the sequence window is expired.

protolambda commented on specs/derivation.md:
```diff
@@ -490,27 +491,63 @@ Note that the presence of any gaps in the batches derived from L1 means that thi
+  within time will be dropped, in favor of an empty batch that can advance the L1 origin.
+- `batch.transactions`: `drop` if the `batch.transactions` list contains a transaction
+  that is invalid or derived by other means exclusively:
+  - any transaction that is empty (zero length byte string)
```
Ignoring an empty tx bytestring is required since we at least need 1 byte to check the tx type.

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,132 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+	// We filter in-place by only remembering the batches that may be processed in the future, or those we are undecided on.
+	remaining := bq.batches[:0]
+batchLoop:
+	for i, batch := range bq.batches {
```
Changing this to be grouped by timestamp :+1:

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,132 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+
+	// Go over all batches, in order of inclusion, and find the first batch we can accept.
+	// We filter in-place by only remembering the batches that may be processed in the future, or those we are undecided on.
+	remaining := bq.batches[:0]
```
Changing this anyway as part of the group timestamp thing, so it won't be a problem. Although it's somewhat tricky, I do think in-place filtering like this is an undervalued Go pattern.

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,132 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
-	// Note that we expect the number of batches for the same timestamp to be small (frequently just 1 ).
-	for _, batch := range batches {
-		l1OriginTime := bq.l1Blocks[0].Time
+	if nextBatch == nil {
```
Will un-nest the `nextBatch==nil` case so it's more readable :+1:

protolambda commented on specs/derivation.md:
```diff
@@ -490,27 +491,63 @@ Note that the presence of any gaps in the batches derived from L1 means that thi
+- `undecided`: we are lacking L1 information until we can proceed batch filtering.
+- `future`: the batch may be valid, but cannot be processed yet and should be checked again later.
+
+The batches are processed in order of the inclusion on L1: if multiple batches can be `accept`-ed the first is applied.
```
Clarified the usage of the rules is in the given order :+1:

protolambda commented:
- Rebased onto `develop`
- Implemented the grouping of batches by timestamp, to handle the performance/DoS case, otherwise still in inclusion order, clarified in specs that future batches can be deferred in processing since these are marked as `future` anyway.
- Clarified a bunch of comments/code based on review, changed structure a tiny bit to make it more readable

Ready for final review/merge.

tynes commented on op-node/rollup/derive/attributes_queue.go:
```diff
@@ -63,9 +64,14 @@ func (aq *AttributesQueue) Step(ctx context.Context, outer Progress) error {
+	safeL2Head := aq.next.SafeL2Head()
+	// sanity check parent hash
+	if batch.ParentHash != safeL2Head.Hash {
+		return NewCriticalError(fmt.Errorf("valid batch has bad parent hash %s, expected %s", batch.ParentHash, safeL2Head.Hash))
```
What exactly does a `CriticalError` indicate? Could a malicious sequencer use this line to dos verifiers?

protolambda replied:
It indicates that there's some programming error. This should never happen: the batch should never be "future" if we're looking up the batch candidates for the exact timestamp we need.

tynes commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+			nextBatch = batch
+			// don't keep the current batch in the remaining items since we are processing it now,
+			// but retain every batch we didn't get to yet.
+			remaining = append(remaining, candidates[i+1:]...)
```
Can this go out of bounds on the last element?

protolambda replied:
The last element is index `len(candidates)-1`, adding `1` makes it `len(candidates)` and `candidates[len(candidates):]` is a valid slice of 0 length.

tynes commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
-			bq.log.Info("found eager batch", "batch", next.Batch)
-			return []*BatchData{next.Batch}, nil
+		// sanity check
+		if nextBatch.Batch.EpochNum > rollup.Epoch(epoch.Number)+1 {
```
This check could be more strict

```
if nextBatch.Batch.EpochNum != rollup.Epoch(epoch.Number)
  && nextBatch.Batch.EpochNum != rollup.Epoch(epoch.Number)+1
```

This would silently fail for going backwards (which I assume isn't allowed). Could also just add another `if` clause that explicitly checks the going backwards case

protolambda replied:
`CheckBatch` already checks these edge-cases. What do you think about removing the sanity check here? It was mostly a product of testing/debugging

tynes replied:
I'm cool with removing this sanity check if its already checked elsewhere

tynes commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
-	if l1BlockRef.Hash != batch.Batch.EpochHash {
-		bq.log.Debug("Batch epoch hash does not match expected L1 block hash", "batch_epoch", batch.Batch.Epoch(), "expected", l1BlockRef.ID())
-		return false, nil
+	validity := CheckBatch(bq.config, bq.log, bq.l1Blocks, bq.next.SafeL2Head(), &data)
```
Passing a logger into `CheckBatch` feels kinda weird to me but I get it, why use `bq.log` here instead of the defined `log` above?

protolambda replied:
`CheckBatch` already adds the same logging details, so that the details are there for every caller.

protolambda replied:
I'll try clean up the logging a bit

protolambda replied:
Removed the logging there, in favor of making the `CheckBatch` logging, which covers all cases now, and has warn-level if the batch needs to be dropped.

tynes commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+func (bq *BatchQueue) deriveNextBatch(ctx context.Context) (*BatchData, error) {
 	if len(bq.l1Blocks) == 0 {
-		return nil, io.EOF
+		panic("cannot derive next batch, no origin was prepared")
```
Are you sure a `panic` in the hot code path is a good idea? Just worried about a bug triggering this since there is a bunch of slicing of `bq.l1Blocks`. Could it return a scary error instead?

protolambda replied:
Will change it to a scary `NewCriticalError`

tynes replied:
This one looks to be still not resolved

protolambda replied:
Ah, good catch

protolambda replied:
Confused it with a different panic that needed to be fixed in the same way

tynes commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
+			bq.batches[nextTimestamp] = remaining
+			return nil, io.EOF
+		default:
+			panic(fmt.Errorf("unknown batch validity type: %d", validity))
```
Exhaustive pattern matching is a nice feature to have in a language, the compiler would force this to be unreachable

protolambda replied:
Rust vs Go :smile:

tynes commented on op-node/rollup/derive/batches.go:
```diff
@@ -1,106 +1,124 @@
-	return
-}
+	// add details to the log
+	log = log.New(
```
Is this additive in its properties or does it overwrite them? In one case, the caller passes in a logger that already has details, in another case it does not. I think the caller should be responsible for adding details to the logger before passing in

protolambda replied:
It applies the latest properties, without mutating the logger of the caller. It copies copies the context values to a new slice within `New`, and concatenates them all with the old ones. Up to the log handler technically to pick the last context values if there are duplicates.

protolambda replied:
See updates, de-duplicated this log context data.

tynes commented on op-node/rollup/derive/batches.go:
```diff
@@ -1,106 +1,124 @@
-		if txBytes[0] == types.DepositTxType {
-			return fmt.Errorf("sequencers may not embed any deposits into batch data, but tx %d has one", i)
+
+	epoch := l1Blocks[0]
```
Perhaps move this declaration to the top of the function after checking the length and use it instead of referencing `l1Blocks[0]`

tynes commented on op-node/rollup/derive/batches.go:
```diff
@@ -1,106 +1,124 @@
+	epoch := l1Blocks[0]
+
+	// Check the L1 origin of the batch
+	batchOrigin := epoch
```
Why this extra intermediate variable `batchOrigin`? I don't see it changed

protolambda commented on specs/derivation.md:
```diff
@@ -392,14 +392,15 @@ Recall that a batch contains a list of transactions to be included in a specific
+| 0               | `rlp_encode([parent_hash, epoch_number, epoch_hash, timestamp, transaction_list])` |
 
 where:
 
```
Yep, same as transaction-typing, will clarify the specs :+1:

norswap reviewed (approved):
Approving to move things along, flagging two things:

- We need to have a conversation about batch replayability (will create linear ticket).
- Still have my concerns with the state-machine-ish spec structures, but this can be easily addressed when doing l2 chain derivation refactoring work, so perfectly fine for now.
     - This will conflict with my last derivation spec PR, what I'll do there is keep both version side by side (old version just need an extra constraint for the parent hash), and we'll consolidate as part of refactors.
